packages/agent/src/polling/index.ts

```diff
@@ -1,53 +1,154 @@
-import { Principal } from '@dfinity/principal';
-import { Agent, RequestStatusResponseStatus } from '../agent';
-import { Certificate, CreateCertificateOptions, lookupResultToBuffer } from '../certificate';
 import { RequestId } from '../request_id';
-import { toHex } from '../utils/buffer';
+import { strToUtf8, toHex } from '../utils/buffer';
+import { CreateCertificateOptions, Certificate, lookupResultToBuffer } from '../certificate';
+import { Agent, ReadStateResponse } from '../agent/api';
+import { Principal } from '@dfinity/principal';
 
 export * as strategy from './strategy';
 import { defaultStrategy } from './strategy';
+import { ReadRequestType, ReadStateRequest } from '../agent/http/types';
+import { CreateReadStateRequestError } from '../errors';
+import { RequestStatusResponseStatus } from '../agent';
 export { defaultStrategy } from './strategy';
+
 export type PollStrategy = (
   canisterId: Principal,
   requestId: RequestId,
   status: RequestStatusResponseStatus,
 ) => Promise<void>;
+
 export type PollStrategyFactory = () => PollStrategy;
 
+interface SignedReadStateRequestWithExpiry extends ReadStateRequest {
+  body: {
+    content: Pick<ReadStateRequest, 'request_type' | 'ingress_expiry'>;
+  };
+}
+
+/**
+ * Options for controlling polling behavior
+ */
+export interface PollingOptions {
+  /**
+   * A polling strategy that dictates how much and often we should poll the
+   * read_state endpoint to get the result of an update call.
+   * @default defaultStrategy()
+   */
+  strategy?: PollStrategy;
+
+  /**
+   * Whether to reuse the same signed request for polling or create a new unsigned request each time.
+   * @default false
+   */
+  preSignReadStateRequest?: boolean;
+
+  /**
+   * Optional replacement function that verifies the BLS signature of a certificate.
+   */
+  blsVerify?: CreateCertificateOptions['blsVerify'];
+
+  /**
+   * The request to use for polling. If not provided, a new request will be created.
+   * This is only used if `preSignReadStateRequest` is set to false.
+   */
+  request?: ReadStateRequest;
+}
+
+export const DEFAULT_POLLING_OPTIONS: PollingOptions = {
+  strategy: defaultStrategy(),
+  preSignReadStateRequest: false,
+};
+
+/**
+ * Check if an object has a property
+ * @param value the object that might have the property
+ * @param property the key of property we're looking for
+ */
+function hasProperty<O extends object, P extends string>(
+  value: O,
+  property: P,
+): value is O & Record<P, unknown> {
+  return Object.prototype.hasOwnProperty.call(value, property);
+}
+
+/**
+ * Check if value is a signed read state request with expiry
+ * @param value to check
+ */
+function isSignedReadStateRequestWithExpiry(
+  value: unknown,
+): value is SignedReadStateRequestWithExpiry {
+  function isObjectWithProperty<O extends object, P extends string>(
```

Comment:
This is much better now, thanks! I would just now rather move `isObjectWithProperty` and `hasFunction` outside the `isSignedReadStateRequestWithExpiry` function, especially because they are so generic and could be used elsewhere.